Improve search/indexing when dealing with URIs

werner291 opened the pull request:
### Pull Request Checklist

<!-- Please read CONTRIBUTING.rst before submitting your pull request -->

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst#changelog)
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst#sign-off)

This should solve #3024 

Tested through a case in the test suite and by deploying the commit on my homeserver. Searching for even small parts of the URI appear to work splendidly,

richvdh commented on synapse/storage/search.py:
```diff
@@ -334,7 +335,8 @@ def store_search_entries_txn(self, txn, entries):
                     entry.key,
-                    entry.value,
+                    # Only store words while discarding special characters
+                    ' '.join(re.findall(r"([\w\-]+)", entry.value, re.UNICODE)),
```
1. what exactly is this doing? I suggest writing a longer comment outside the `args = ` expression, giving some examples and explanation.

2. `re.UNICODE` is obsolete

3. Any reason we use `re.findall` rather than `re.finditer`?

werner291 replied:
Does the same thing as is done here: https://github.com/matrix-org/synapse/blob/develop/synapse/storage/search.py#L695

No particular reason, `finditer` should work too.

I'll add some comments with explanations.

werner291 replied:
Not sure what to do about the UNICODE thing, can we just remove it?

richvdh commented on changelog.d/3024.bugfix:
```diff
@@ -0,0 +1 @@
+Improved search performance when dealing with cases such as URIs in messages.
```
```suggestion
Improve search performance when dealing with cases such as URIs in messages.
```

Feel free to add a `Contributed by @werner291.` to get your name in the changelog :)

werner291 commented:
I've refactored the code such that the keyword extraction is now shared between the indexing and search query processing to ensure uniformity between the two.

I've also added some explanation comments.

ara4n commented on synapse/storage/search.py:
```diff
@@ -322,19 +322,29 @@ def store_search_entries_txn(self, txn, entries):
             )
 
+            # About the word extraction: PostgreSQL appears to be optimized for words,
+            # so it chokes on code and werd formatting such as URIs.
```
```suggestion
            # so it chokes on code and weird formatting such as URIs.
```

ara4n commented:
thanks for looking at this! one thought: presumably we should somehow trigger a reindex in order to pull in the fix?

werner291 commented:
> thanks for looking at this! one thought: presumably we should somehow trigger a reindex in order to pull in the fix?

A re-index would be beneficial to make the improved search work for older messages, indeed. If we don't, older messages will simply be found with the same performance as before.

richvdh commented on synapse/storage/search.py:
```diff
@@ -322,19 +322,29 @@ def store_search_entries_txn(self, txn, entries):
+            # so it chokes on code and weird formatting such as URIs.
+            # This improves search indexing by only picking up the words.
+            # For example: "https://github.com/matrix-org/synapse/pull/6132"
+            # Becomes "https github com matrix org synapse pull 6132"
```
```suggestion
            # becomes "https github com matrix org synapse pull 6132"
```

richvdh commented on synapse/storage/search.py:
```diff
@@ -683,6 +693,17 @@ def f(txn):
+    Extract an iterator of words from the string.
+
+    For instance: https://github.com/matrix-org/synapse/issues/3024
+    Becomes: ["https","github","com","matrix-org","synapse","issues","3024"]
```
```suggestion
    becomes: ["https","github","com","matrix-org","synapse","issues","3024"]
```

richvdh commented on synapse/storage/search.py:
```diff
@@ -683,6 +693,17 @@ def f(txn):
+    :param entry: A string to be pre-processed for searching.
+    :return: List of words in the string.
+    """
+    return re.finditer(r"([\w\-]+)", entry.value, re.UNICODE)
```
re.UNICODE is a no-op:

```suggestion
    return re.finditer(r"([\w\-]+)", entry.value)
```

richvdh commented on changelog.d/3024.bugfix:
```diff
@@ -0,0 +1 @@
+Improve search performance when dealing with cases such as URIs in messages.
```
```suggestion
Improve search performance when dealing with cases such as URIs in messages. Contributed by @werner291.
```

richvdh commented on synapse/storage/data_stores/main/search.py:
```diff
@@ -693,6 +704,21 @@ def _to_postgres_options(options_dict):
+
+    Returns: the words in the string.
+    """
+    return re.finditer(r"([\w\-]+)", entry.value)
```
wait, if `entry` is a str, it doesn't have a `value` property.

Also: `finditer` returns `Match` objects, not `str`s. You'll either need to extract the matched string from the Match objects (`m[0] for m in matches`), or go back to `findall`.

richvdh commented:
Afraid I'm going to close this because it seems to be bitrotting.
